layer2-state-provider/history_commitment_provider.go

```diff
@@ -0,0 +1,281 @@
+		if end > maxHeightForLevel {
+			return 0, fmt.Errorf(
+				"end %d was greater than max height for level %d",
+				end,
+				maxHeightForLevel,
+			)
+		}
+	}
+	if end < start {
+		return 0, fmt.Errorf("invalid range: end %d was < start %d", end, start)
+	}
+	// The number of hashes is the difference between the start and end
+	// requested heights, plus 1.
+	return uint64(end-start) + 1, nil
+}
+
+// Figures out the actual opcode index we should move the machine to
+// when we compute the history commitment. As there are different levels of challenge
+// granularity, we have to do some math to figure out the correct index.
+//
+// Take, for example, that we have 5 challenge kinds:
+//
+// block    => over a range of L2 message hashes
+// colossal => over a range of giant steps
+// giant    => over a range of big steps
+// big      => over a range of small steps
+// small    => over a range of individual WAVM opcodes
+//
+// We only directly step through WAVM machines when in a subchallenge (starting at colossal),
+// so we can ignore block challenges for this calculation.
+//
+// Next, take the following constants:
+//
+//	colossal_steps_per_giant_step = 16
+//	giant_steps_per_big_step      = 8
+//	big_steps_per_giant_step      = 4
+//	small_steps_per_big_step      = 2
+//
+// Let's say we want to figure out the machine start opcode index for the following inputs:
+//
+//	colossal_step=4, giant_step=5, big_step=1, small_step=0
+//
+// We can compute the opcode index using the following algorithm for the example above.
+//
+//	  4 * (8 * 4 * 2)
+//	+ 5 * (4 * 2)
+//	+ 1 * (2)
+//	+ 0
+//	= 298
+//
+// This generalizes for any number of subchallenge levels into the algorithm below.
+// It works by taking the sum of (each input * product of all challenge level height constants beneath its level).
+// For the block challenge level, the machine start opcode index is always 0.
+func (p *HistoryCommitmentProvider) computeMachineStartIndex(
+	startHeights validatedStartHeights,
+) OpcodeIndex {
+	if len(startHeights) == 1 {
+		return 0
+	}
+	// We ignore the block challenge level here.
+	heights := startHeights[1:]
+	leafHeights := p.challengeLeafHeights[1:]
+
+	// Next, we compute the opcode index.
+	opcodeIndex := uint64(0)
+	idx := 1
+	for _, height := range heights {
+		total := uint64(1)
+		for i := idx; i < len(leafHeights); i++ {
+			total *= uint64(leafHeights[i])
+		}
+		opcodeIndex += total * uint64(height)
+		idx += 1
+	}
+	return OpcodeIndex(opcodeIndex)
+}
+
+// Computes the step size for stepping through a machine at a block height. Each challenge level has a different
+// amount of steps, so the total step size can be computed as a multiplication of all the
+// next challenge levels needed.
+func (p *HistoryCommitmentProvider) computeStepSize(startHeights validatedStartHeights) (StepSize, error) {
+	challengeLevel := len(startHeights) - 1
+	totalChallengeLevels := len(p.challengeLeafHeights)
```

Comment:
This function is based around protocol constants, not on user input, so that shouldn't apply. What this tries to answer is: "How many individual opcodes are there in a single step of a Megastep challenge?" and the answer is 1 million